# Huawei Research France

## Transfer learning on home network: 

## Build a transfer learning solution for home network failure prediction

_Balázs Kégl, Aladin Virmaux, Illyyne Saffar, Jianfeng Zhang (Huawei Research, Noah's Ark Laboratory, France)_

## Introduction

Optical access network is a main-stream Home Broadband Access Network solution around the world. It connects terminal subscribers to their service provider. When the network suffers a failure, an immense loss of
data occurs, and affects hardly both of the quality of the service (QoS) and the user's experience (the quality of Experience QoE). To handle such issue and reduce the caused damage, it is important to predict in advance the network failures and fix them in time. Machine learning (ML) algoritms has been wildely used as a solution to build these failure prediction models. However, most of ML models are data-specific and are prone to degradation when the data distribution changes. This year's Huawei France hackathon aims to solve this problem. You will receive a labeled optical access network dataset from a city A (which we termed it as _source_ domain) and a mostly unlabeled  dataset (only 20% of labels) from a city B (which we termed it as _target_ domain). You are asked to build a transfer learning solution using the labaled source plus the unlabeled target to train a failure prediction model for city B. It is an **unsupervised domain adaptation** problem. An additional challenge can be faced from:

1. **Missing values**: There are a lot of missing values in the data.
2. **Time series sensor data**: Observation or samples are not mutually independent.
3. **Class imbalance**: Network failure is a rare event, thus it is very imbalanced classification problem. 
4. **Scoring metrics**: Some of them are a nonstandard measures.


## Context
Transmission technologies have evolved to integrate optical technologies even in access networks, as close as possible to the subscriber. Due to its ability to propagate over long distances without signal regeneration, its low latency, and its very high bandwidth, fiber optic networks are so far the transmission medium par excellence. Optical fiber, initially deployed in very long distance and very high speed networks, is now tending to be generalized to offer more consuming services in terms of bandwidth. These are FTTH technologies for Fiber to the Home.

The FTTH architecture generally adopted by operators is a PON (Passive Optical Network) architecture. The PON is a point-to-multipoint architecture based on the following elements:
- A shared fiber optic infrastructure. The use of optical couplers in the network is the basis of the architecture and deployment engineering. The couplers are used to serve several zones or several subscribers.
- A center equipment acting as Optical Line Termination (OLT). The OLT manages the broadcasting and reception of streams through network interfaces. It receives signals from subscribers and broadcasts a content based on specific services. 
- An end equipments.
    - ONT (Optical Network Terminations) in case where the equipment is dedicated to a customer and the fiber reaches the customer. This is then an FTTH (Fiber To The Home) type architecture. There is only one fiber per customer (signals are bidirectional)
    - ONU (optical network unit) in the case where the equipment is dedicated to a whole building. This is then an FTTB (Fiber To The Building).
    
<img src="https://image.makewebeasy.net/makeweb/0/p4Ky6EVg4/optical%20fiber-knowledge/Apps_FTTx_Fig3.png">

The data for this challenges is gathered from sensors collected in ONT level. 

### The data

The data is coming from two different cities: city A (the source) and city B (the target). Data is labled for town A but unlabled for B (only 20% of labled data is known for city B). For both cities A and B data is a time serie gathered along a total of around 60 days. The time series granularity is 15 minutes. The samples reprensent different users (thus different ONT). At each time step, we have a 10 dimensional measurement of the following parameters (in parenthesis are the units of each feature).
- data/ts/: 
  - current: bias current of the GPON ONT optical module (mA)
  - err_down_bip: number of ONT downstream frames with BIP error (integer)
  - err_up_bip: number of ONT upstream frames with BIP error (integer)
  - olt_recv: receiving power of the GPON ONT optical module from the ONU (dBm)
  - rdown: downstream rate of GPON ONT (Mbs)
  - recv: receiving power of the GPON ONT optical module (dBm)
  - rup: upstream rate of GPON ONT (Mbs)
  - send: transmitting power of the GPON ONT optical module (dBm)
  - temp: temperature of the GPON ONT optical module (Celsius)
  - volt: power feed voltage of GPON ONT optical module (mV)
- data/labels/: -1 (all good), 0 (weak) or 1 (failure) for sample. Let $x_t$ be the sample collected at the day $t$, then the label corresponding is computed on the day $t+7$. We aim to predict a failure from data coming 7 days before.  


The data is given to you with shape **[users, timestamps, features]** and the features are given in the same orders as presented above.

#### Missing data

You will notice that a some data is missing in the datasets. There may be several reaons:

1. no data was gathered on a specific date for a specific user.
2. the data collecting process fail to retrieve a feature.
    
It is part of the challenge to overcome this real-life difficulty.

### The scoring metrics

In this challenge we propose to evaluate the performance using 5 different metrics:

- **Acc** Accuracy: The number of truely predicted labels over the total number of the samples [sklearn function](https://scikit-learn.org/stable/modules/generated/sklearn.metrics.accuracy_score.html#sklearn.metrics.accuracy_score). 
- **Auc** roc_auc score: AUC measures the entire two-dimensional area underneath the ROC curve. This score gives us a good idea of how well the classifier will perform [sklearn function](https://scikit-learn.org/stable/modules/generated/sklearn.metrics.roc_auc_score.html).
- **AP** Average precision: AP summarizes a precision-recall curve as the weighted mean of precisions achieved at each threshold, with the increase in recall from the previous threshold used as the weight: $\sum_n (R_n - R_{n-1}) P_n$ where $P_n$ and $R_n$ are the precision and the recall at the $n$-th treshold [sklearn function](https://scikit-learn.org/stable/modules/generated/sklearn.metrics.average_precision_score.html#sklearn.metrics.average_precision_score).
- **Precision@Recall**: It is a hybrid score implemented in `utils.scores` and it computes the precision when the recall is at some percentage, that is to say: 
  - rec@5: Recall at 5% computes the precision when the recall is at 5%.
  - rec@10: Recall at 10% computes the precision when the recall is at 10%

Note that the AP is the metric used for the final evaluation.


## Competition rules

The rules may be adjusted before the start of the challenge, June 16 2021.

[//]: # "* Submissions will be trained on a time series of roughly 50k samples with a  time window length of 672, across 10 features  and tested on a time series of roughly 35k samples." 
* The competition will end on June 27, 2020 at 18h UTC (20h in Paris).
* The competitive phase will be followed by a collaborative phase when participants may (and encouraged to) reuse and combine each other's code.
* The final results will be announced at the closing ceremony on June 30. We will also ask top teams to present their approaches at the event.
* All models will be trained on the same cloud server allowing 76 CPUs (with shared memory of 500GB RAM).
* Participants will be given a total of 20 machine hours (per cross-validation fold). Submissions of a given participant will be ordered by submission timestamp. We will make an attempt to train all submissions, but starting from (and including) the first submission that makes the participant's total training time exceed 20 hours, all submissions will be disqualified from the competition (but can enter into the collaborative phase). Testing time will not count towards the limit. Training time will be displayed on the leaderboard for all submissions, rounded to second. If a submission raises an exception, its training time will not count towards the total.
* There is a timeout of 1 day between submissions that did not raise an exception.
* Submissions submitted after the end of the competition will not qualify for prizes.
* The public leaderboard will display validation scores running a cross-validation. The official scores will be calculated on the hidden test set and will be published after the closing of the competition. We will rank submissions according to their AP score.
* The organizers will do their best so that the provided backend runs flawlessly. We will communicate with participants in case of concerns and will try to resolve all issues, but we reserve the right to make unilateral decisions in specific cases, not covered by this set of minimal rules.
* The organizers reserve the right to disqualify any participant found to violate the fair competitive spirit of the challenge. Possible reasons, without being exhaustive, are multiple accounts, attempts to access the test data, etc.
* Participants can form teams outside the platform before submitting any model individually, and submit on a single team account. Participating in more than one team at the same time is against the "no multiple accounts" rule, so, if discovered, may lead to disqualification. Before signing up, teams should communicate their composition and team name to BeMyApp.
* Participants retain copyright on their submitted code and grant reuse under BSD 3-Clause License.

Participants accept these rules automatically when making a submission at the RAMP site.

## Getting started


Besides the usual pydata libraries, you will need to install `ramp-workflow` from the advanced branch:
```
pip install git+https://github.com/paris-saclay-cds/ramp-workflow.git@advanced
```


It will install the `rampwf` library and the `ramp-test` script that you can use to check your submission before submitting. You do not need to know this package for participating in the challenge, but it could be useful to take a look at the [documentation](https://paris-saclay-cds.github.io/ramp-docs/ramp-workflow/advanced/index.html) if you would like to know what happens when we test your model, especially the [RAMP execution](https://paris-saclay-cds.github.io/ramp-docs/ramp-workflow/advanced/scoring.html) page to understand `ramp-test`, and the [commands](https://paris-saclay-cds.github.io/ramp-docs/ramp-workflow/advanced/command_line.html) to understand the different command line options. 

In [1]:
import numpy as np
import rampwf as rw

Read `problem.py` so you can have an access to the same interface as the testing script.

In [2]:
problem = rw.utils.assert_read_problem()

### The data

First take the public data set from the .... #optical_network_challenge channel (join by [clicking here](link)) and unzip it to create `data/`. Note that the public data is different from the private one. 

The train data is composed of source and target data coming respectively from city A and city B. In real life FTTH problem presents 3 classes: 1) the case where the flow is normal and everything is going smoothly, 2) the case where the flow is poor but the conexion still working and 3) the case where we face a failure. Thus we can set 3 classes \[good, poor, failure\]. For the OAN failure detection we are interested in a binary classification between the two classes \[poor, failure\]. You are free to exploit the data of the \[Good\] class but in the scoring you are only judged on the binary classification. 

The dataset you are given is composed of:
- The source data is composed of `X_train.source` and `X_train.source_bkg`. The labels of the source are accessible in `y_train.source`.
    - `X_train.source`: Data for the classes \[poor, failure\]
    - `X_train.source_bkg`: Data for the class \[Good\] 
    - `y_train.source`: Labels for the `X_train.source`, where 0: poor and 1 failure.
- The target is composed of `X_train.target_labeled`, `X_train.target_unlabeled` and `X_train.target_bkg`. The labels of the target are accessible in `y_train.target`. 
    - `X_train.target_labeled`: Target labeled data for the classes \[poor, failure\]
    - `X_train.target_unlabeled`: Target unlabeled data
    - `X_train.target_bkg`: Target data for the class \[Good\] 
    - `y_train.target`: Labels for the `X_train.target_labeled`, where 0: poor and 1 failure.

Read the training and test data.

In [3]:
X_train, y_train = problem.get_train_data()
X_test, y_test = problem.get_test_data()

Train data
Optical Dataset composed of
46110 source samples
50862 source background samples
438 target labeled samples
8202 target unlabeled samples
29592 target background samples
 Optical Dataset labels composed of
46110 labels of source samples
438 labels of target samples

Test data
Optical Dataset composed of
0 source samples
0 source background samples
17758 target labeled samples
0 target unlabeled samples
47275 target background samples
 Optical Dataset labels composed of
0 labels of source samples
17758 labels of target samples



The input data is a time series of 10 features and 672 time length. But, as you can see in the cells bellow, it contains nan values thus it should be cleaned.

In [4]:
X_train.source[7]

array([[1.30e+01, 0.00e+00, 0.00e+00, ..., 2.28e+00, 4.40e+01, 3.26e+03],
       [1.30e+01, 0.00e+00, 0.00e+00, ..., 2.23e+00, 4.40e+01, 3.26e+03],
       [1.30e+01, 0.00e+00, 0.00e+00, ..., 2.31e+00, 4.40e+01, 3.26e+03],
       ...,
       [     nan,      nan,      nan, ...,      nan,      nan,      nan],
       [     nan,      nan,      nan, ...,      nan,      nan,      nan],
       [     nan,      nan,      nan, ...,      nan,      nan,      nan]])

In [5]:
X_train.source[10000]

array([[1.4000e+01, 5.3560e+03, 0.0000e+00, ..., 2.1500e+00, 4.7000e+01,
        3.3000e+03],
       [1.4000e+01, 6.2650e+03, 2.0000e+00, ..., 2.0800e+00, 4.7000e+01,
        3.3200e+03],
       [1.4000e+01, 7.8850e+03, 4.0000e+00, ..., 2.4600e+00, 4.7000e+01,
        3.3000e+03],
       ...,
       [1.4000e+01, 5.1556e+04, 6.0000e+00, ..., 2.3000e+00, 4.6000e+01,
        3.3200e+03],
       [1.4000e+01, 4.3742e+04, 2.0000e+01, ..., 1.9500e+00, 4.6000e+01,
        3.3000e+03],
       [1.4000e+01, 4.4794e+04, 2.6000e+01, ..., 2.2800e+00, 4.6000e+01,
        3.3000e+03]])

### The classification task

You can load here the naive domain adaptation implemented in the one of the folders named `starting_kit`. It is a naive transfer where the model trained on the source to classify the target using random forest.

In [6]:
# %load submissions/starting_kit/classifier.py
from sklearn.ensemble import RandomForestClassifier
from utils.dataset import OpticalDataset, OpticalLabels

import numpy as np

class Classifier:

    def __init__(self):
        self.clf = RandomForestClassifier(
            n_estimators=2, max_depth=2, random_state=44, n_jobs=-1)

    def fit(self, X_source, X_source_bkg, X_target, X_target_unlabeled,
            X_target_bkg, y_source, y_target):
        self.clf.fit(X_source, y_source)

    def predict_proba(self, X_target, X_target_bkg):
        y_proba = self.clf.predict_proba(X_target)
        return y_proba


You can train your submission and obtain test predictions using the same protocol as our training script.

But first you will need to flat the data to run the naive transfer.

In [7]:
trained_workflow = problem.workflow.train_submission('submissions/starting_kit', X_train, y_train)
y_test_pred = problem.workflow.test_submission(trained_workflow, X_test)

### The scores

We compute 5 scores on the classification. All scores are implemented in `external_imports.utils.scores.py` so you can look at the precise definitions there.
**The official score of the competition is AP.**

In [8]:
AP    = problem.score_types[0]
rec5  = problem.score_types[1]
rec10 = problem.score_types[2]
rec20 = problem.score_types[3]
acc   = problem.score_types[4]
auc   = problem.score_types[5]

In [9]:
print('AP test score    = {}'.format(AP(y_test.target, y_test_pred[:,1])))
print('rec5 test score  = {}'.format(rec5(y_test.target, y_test_pred[:,1])))
print('rec10 test score = {}'.format(rec10(y_test.target, y_test_pred[:,1])))
print('rec20 test score = {}'.format(rec20(y_test.target, y_test_pred[:,1])))
print('acc test score   = {}'.format(acc(y_test.target, y_test_pred.argmax(axis=1))))
print('auc test score   = {}'.format(auc(y_test.target, y_test_pred[:,1])))

AP test score    = 0.1626234044092243
rec5 test score  = 0.07541412380122058
rec10 test score = 0.1970357454228422
rec20 test score = 0.34132519616390583
acc test score   = 0.821939407590945
auc test score   = 0.586923967966097


### The cross validation scheme

The cross-validation follows the same scheme as the train/test cut (see `problem.get_cv`): a 10 train/test cross validation folders across both the source and the target is  proposed for `problem.get_cv`. To learn more about these cuts details, you can check `external_imports.utils.cv.py` and then the `TLShuffleSplit` class.

You are free to play with both the train/test cut and the cross-validation when developing your models but be aware that we will use the same set up on the official server as the one in the RAMP kit (on a different set of four campaigns that will not be available to you).

The following cell goes through the same steps as the official evaluation script (`ramp-test`).

In [10]:
splits = problem.get_cv(X_train, y_train)

y_test_preds = []
for fold_i, (train_is, valid_is) in enumerate(splits):
    trained_workflow = problem.workflow.train_submission(
        'submissions/starting_kit', X_train, y_train, train_is)
    X_fold_train = X_train.slice(train_is)
    X_fold_valid = X_train.slice(valid_is)
    
    y_train_pred = problem.workflow.test_submission(trained_workflow, X_fold_train)
    y_valid_pred = problem.workflow.test_submission(trained_workflow, X_fold_valid)
    y_test_pred = problem.workflow.test_submission(trained_workflow, X_test)
    print('-------------------------------------')
    print('training AP on fold {} = {}'.format(fold_i, AP(y_train.slice(train_is).target, y_train_pred[:,1])))
    print('validation AP on fold {} = {}'.format(fold_i, AP(y_train.slice(valid_is).target, y_valid_pred[:,1])))
    print('test AP on fold {} = {}'.format(fold_i, AP(y_test.target, y_test_pred[:,1])))
    
    y_test_preds.append(y_test_pred)

-------------------------------------
training AP on fold 0 = 0.30833333333333335
validation AP on fold 0 = 0.2637875964895809
test AP on fold 0 = 0.16218430339780684
-------------------------------------
training AP on fold 1 = 0.21250000000000002
validation AP on fold 1 = 0.2555942077788053
test AP on fold 1 = 0.16361016472786805
-------------------------------------
training AP on fold 2 = 0.2
validation AP on fold 2 = 0.29440601825201235
test AP on fold 2 = 0.1745388926023523
-------------------------------------
training AP on fold 3 = 0.7375
validation AP on fold 3 = 0.28218715512682335
test AP on fold 3 = 0.16904411795376056
-------------------------------------
training AP on fold 4 = 0.21250000000000002
validation AP on fold 4 = 0.24879604051634688
test AP on fold 4 = 0.16172210972525408
-------------------------------------
training AP on fold 5 = 0.4
validation AP on fold 5 = 0.30569665610952207
test AP on fold 5 = 0.16752441466614315
-------------------------------------
tr

We compute both the mean test score and the score of bagging your eight models. The official ranking will be determined by the bagged test score (on different data sets from the ones you have). Your public score will be the bagged validation score (the averaging is [slightly more complicated](https://github.com/paris-saclay-cds/ramp-workflow/blob/master/rampwf/utils/combine.py#L56) since we need to take care of the cross validation masks properly). 

In [11]:
bagged_y_pred = np.array(y_test_preds).mean(axis=0)
print('Mean AP score = {}'.format(
    np.mean([AP(y_test.target, y_test_pred[:,1]) for y_test_pred in y_test_preds])))
print('Bagged AP score = {}'.format(
    AP(y_test.target, np.array([y_test_pred for y_test_pred in y_test_preds]).mean(axis=0)[:,1])))

Mean AP score = 0.1670974152623759
Bagged AP score = 0.16968309345277285


## Example submissions

Besides the starting kit implementing a naive transfer classifier using random forest, we provide you an other simple naive transfer example using xgboost, to get you started. 

Note that those submissions are not fully hyperparameters-optimized and do not come with every possible tricks. Feel free to play with it locally. There is no need to submit them since their respective score are already in the
leaderboard.


### Data cleaning/ reshaping
The data also needs to be cleaned proprely since it contains nan values. With every submission you find, besides of the `classifier.py`, a `feature_extractor.py`, where you can pre_process the data. For eg, in both `starting_kit` and `source_rf` or `target_rf` a data reshaping and a nan values handling are proposed in `FeatureExtractor.transform`:

In [12]:
def transform(self, X):
    # Deal with NaNs inplace
    np.nan_to_num(X, copy=False)
    # We flatten the input, originally 3D (sample, time, dim) to
    # 2D (sample, time * dim)
    X = X.reshape(X.shape[0], -1)
    return X

You can change the `feature_extractor.py` to exploit or to preprocess more the data. Once again, do not hesitate to play with those examples locally to get familiar with RAMP and the data.

### Source_rf submission
This model is submitted under `submission.source_rf.classifier.py`. This model is very similar to the `starting_kit.classifier.py`. It differes with better the hyperparameters. 

In `source_rf` a random forest classifier is trained on the `X_train.source` data and tested on `X_test`. The oan failure is then detected without considering or exploiting the information in `X_train.source_bkg` or  `X_train.target_labeled`, `X_train.target_unlabeled`, `X_train.target_bkg`. This can be seen as the lower bound of the transfer learning for the oan failure detection. 

In [13]:
# %load submissions/source_rf/classifier.py
from sklearn.ensemble import RandomForestClassifier

class Classifier:

    def __init__(self):
        self.clf = RandomForestClassifier(
            n_estimators=50, max_depth=10, random_state=44, n_jobs=-1)

    def fit(self, X_source, X_source_bkg, X_target, X_target_unlabeled,
            X_target_bkg, y_source, y_target):
        self.clf.fit(X_source, y_source)

    def predict_proba(self, X_target, X_target_bkg):
        y_proba = self.clf.predict_proba(X_target)
        return y_proba

### Target_rf submission

This model is submitted under `submission.target_rf.classifier.py`.  In `target_rf` a random forest classifier is trained on the `X_train.target_labeled` data and tested on `X_test`. The oan failure is then detected without considering or exploiting the information in `X_train.source` or  `X_train.source_bkg`. This can be seen as the upper bound of the transfer learning for the oan failure detection. 

In [14]:
# %load submissions/target_rf/classifier.py
from sklearn.ensemble import RandomForestClassifier

class Classifier:

    def __init__(self):
        self.clf = RandomForestClassifier(
            n_estimators=50, max_depth=10, random_state=44, n_jobs=-1)

    def fit(self, X_source, X_source_bkg, X_target, X_target_unlabeled,
            X_target_bkg, y_source, y_target):
        self.clf.fit(X_target, y_target)


    def predict_proba(self, X_target, X_target_bkg):
        y_proba = self.clf.predict_proba(X_target)
        return y_proba

### Results:
|          | ap             | rec-5         | rec-10         | rec-20         | acc            |  auc           | 
|:---------|:--------------:|:-------------:|:--------------:|:--------------:|:--------------:|:--------------:|   
|source_rf | 0.191 ± 0.0026 | 0.073 ± 0.002 | 0.176 ± 0.0032 | 0.357 ± 0.0075 | 0.84 ± 0.0014  | 0.637 ± 0.0063 | 
|target_rf | 0.163 ± 0.0218 | 0.067 ± 0.0182| 0.138 ± 0.0339 | 0.272 ± 0.0537 | 0.813 ± 0.036  | 0.591 ± 0.0399 | 

## Local testing (before submission)

You submission will contain a single `classifier.py` file implementing a classifier class with a `fit` and `predict_proba` function (scikit-learn API) as in the starting kit. You should place it in the `submission/<submission_name>` folder in your RAMP kit folder. To test your submission, go to your RAMP kit folder in a terminal and type
```
ramp-test --submission <submission_name>
```
It will train and test your submission much like we did it above in this notebook, and print the foldwise and summary scores. You can try it also in this notebook:

In [15]:
!ramp-test --submission target_rf

/bin/bash: ramp-test: command not found


## Submission

1. First you will need to sign up at the [RAMP site](https://ecs-90-84-188-77.compute.prod-cloud-ocb.orange-business.com). Your team will be approved shortly by a system admin who will check that you communicated your team nick and composition to BeMyApp.
2. You will then need a second sign-up, this time for the [optical network challenge](https://ecs-90-84-188-77.compute.prod-cloud-ocb.orange-business.com/events/optical_network_modelling_hackaton). If your site sign up was approved in the previous point, you should see a "Join event" button on the right of the top menu. This request will also be approved by a site admin.
3. Once you are signed up, you can start submitting (once a day). If you are happy with your local scores, copy-paste your submission at the [sandbox](https://ecs-90-84-188-77.compute.prod-cloud-ocb.orange-business.com/events/optical_network_modelling_hackaton/sandbox), press "submit now", name your submission, then give credits to which other submission you used (in the collaborative phase you will see only your own submissions in the list.
4. Your submission will be sent to train. It will either come back with an error or will be scored. You can follow the status at [my submissions](https://ecs-90-84-188-77.compute.prod-cloud-ocb.orange-business.com/events/optical_network_modelling_hackaton/my_submissions).
5. If there is an error, click on the error to see the trace. You can resubmit a failed submission **under the same name**, this will not count in your daily quota.
6. There is no way to delete trained submissions. In exceptional cases we can stop a submission that hasn't been scored yet so you can resubmit. We strongly suggest to finish training at least one fold locally (using `ramp-test`) before submitting so you can estimate the training time.
7. You can follow the scores of the other participants at the [public leaderboard](https://ecs-90-84-188-77.compute.prod-cloud-ocb.orange-business.com/events/optical_network_modelling_hackaton/leaderboard).
8. The public [competition leaderboard](https://ecs-90-84-188-77.compute.prod-cloud-ocb.orange-business.com/events/optical_network_modelling_hackaton/competition_leaderboard) displays the top submission (according to the public score) of each participant. You can change which of your submission enters the competition by pulling out the top submission. Click on the particular submission at [my submissions](https://ecs-90-84-188-77.compute.prod-cloud-ocb.orange-business.com/events/optical_network_modelling_hackaton/my_submissions) and click on the yellow button. The operation is reversible as many times you want, even after the competition deadline.

## Contact

You can contact the organizers in the Slack of the challenge, join by [clicking here](link). 